### PR TITLE
[GAWB-2818] Leonardo: Fix cookie auth expiration using a notebook extension

### PR DESCRIPTION
**Pull request by rtitle**
JIRA: https://broadinstitute.atlassian.net/browse/GAWB-2818
Tech Doc: https://docs.google.com/document/d/1hrxfYpZx5no1eWxzGyyiG4g-_lajaZb8wGqsfsF0qWo/edit?ts=5a03239f

1. Added a google_sign_in.js file which gets plugged into all notebooks and auto-refreshes the Google auth token + cookie while the user is using the notebook.
2. Moved cluster resource files from firecloud-develop to leonardo. See https://github.com/broadinstitute/firecloud-develop/pull/790

Tested in a fiab and it works.

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/resources/jupyter/cluster-docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  image: "${JUPYTER_DOCKER_IMAGE}"
+  net: host
+  volumes:
+    - /work:/home/user/work
+    - /etc/hadoop/conf:/etc/hadoop/conf
+    - /hadoop_gcs_connector_metadata_cache:/hadoop_gcs_connector_metadata_cache
+    - /etc/install-jupyter-extension.sh:/etc/install-jupyter-extension.sh
+    - /etc/${JUPYTER_EXTENSION_ARCHIVE}:/etc/${JUPYTER_EXTENSION_ARCHIVE}
+    - /etc/${JUPYTER_SERVICE_ACCOUNT_CREDENTIALS}:/etc/${JUPYTER_SERVICE_ACCOUNT_CREDENTIALS}
+    - /etc/custom.js:/home/jupyter-user/.jupyter/custom/custom.js
```
Comment (rtitle) [inline src/main/resources/jupyter/cluster-docker-compose.yaml]: Only changes in this file: 
```
+    - /etc/custom.js:/home/jupyter-user/.jupyter/custom/custom.js
+    - /etc/google_sign_in.js:/home/jupyter-user/.jupyter/custom/google_sign_in.js
```

File: src/main/resources/jupyter/cluster-site.conf [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<VirtualHost _default_:${SSL_HTTPD_PORT}>
```
Comment (rtitle) [inline src/main/resources/jupyter/cluster-site.conf]: No change to this file

File: src/main/resources/jupyter/custom.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+define([
+    'base/js/events'
+], function(events) {
```
Comment (rtitle) [inline src/main/resources/jupyter/custom.js]: The entry point to the `custom.js` Jupyter hook. This simply calls `google_sign_in.js`

File: src/main/resources/jupyter/google_sign_in.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    "paths": {
+        "gapi": "https://apis.google.com/js/platform"
+    }
+})
+
+/// TEMPLATED CODE
+var clientId = $(googleClientId);
+///
+
+require(['gapi'], function(gapi) {
```
Comment (rtitle) [inline src/main/resources/jupyter/google_sign_in.js]: See https://docs.google.com/document/d/1hrxfYpZx5no1eWxzGyyiG4g-_lajaZb8wGqsfsF0qWo/edit?ts=5a03239f for some explanation of this code.

File: src/main/resources/jupyter/init-actions.sh [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+JUPYTER_SERVER_CRT=$(jupyterServerCrt)
+JUPYTER_SERVER_KEY=$(jupyterServerKey)
+JUPYTER_ROOT_CA=$(rootCaPem)
+JUPYTER_DOCKER_COMPOSE=$(jupyterDockerCompose)
+JUPYTER_PROXY_SITE_CONF=$(jupyterProxySiteConf)
+JUPYTER_INSTALL_EXTENSION_SCRIPT=$(jupyterInstallExtensionScript)
+JUPYTER_EXTENSION_URI=$(jupyterExtensionUri)
+JUPYTER_SERVICE_ACCOUNT_CREDENTIALS=$(jupyterServiceAccountCredentials)
+JUPYTER_CUSTOM_JS_URI=$(jupyterCustomJsUri)
```
Comment (rtitle) [inline src/main/resources/jupyter/init-actions.sh]: Only changes to this file:
```
+JUPYTER_CUSTOM_JS_URI=$(jupyterCustomJsUri)
+JUPYTER_GOOGLE_SIGN_IN_JS_URI=$(jupyterGoogleSignInJsUri)
```
```
+gsutil cp ${JUPYTER_CUSTOM_JS_URI} /etc
+gsutil cp ${JUPYTER_GOOGLE_SIGN_IN_JS_URI} /etc
```

File: src/main/resources/jupyter/install-jupyter-extension.sh [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (rtitle) [inline src/main/resources/jupyter/install-jupyter-extension.sh]: No changes to this file

File: src/main/resources/jupyter/google_sign_in.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        "gapi": {
+            "exports": "gapi"
+        }
+    },
+    "paths": {
+        "gapi": "https://apis.google.com/js/platform"
+    }
+})
+
+/// TEMPLATED CODE
```
Comment (jmthibault79) [inline src/main/resources/jupyter/google_sign_in.js]: Please add a short description of Leo templating here, to give someone reading this file on its own an idea where to look.

File: src/main/resources/jupyter/google_sign_in.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+                auth2.signIn();
+            }
+        });
+    });
+});
+
+function set_cookie(token, expires_in) {
+    var now = new Date();
+    now.setSeconds(now.getSeconds() + expires_in);
+    document.cookie = "FCtoken="+token+";expires="+now.toUTCString();
```
Comment (jmthibault79) [inline src/main/resources/jupyter/google_sign_in.js]: it's not `now` anymore so I don't like this variable name

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/SwaggerConfig.scala [hunk 9]
```diff
@@ -1,3 +1,5 @@
 package org.broadinstitute.dsde.workbench.leonardo.config
 
-case class SwaggerConfig(googleClientId: String, realm: String)
+import org.broadinstitute.dsde.workbench.leonardo.model.GoogleClientId
+
+case class SwaggerConfig(googleClientId: GoogleClientId, realm: String)
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/SwaggerConfig.scala]: We're still stringly-typing confs, right?

File: src/main/resources/jupyter/google_sign_in.js (hunk 7 above)
Comment (rtitle) [reply src/main/resources/jupyter/google_sign_in.js]: 👍

File: src/main/resources/jupyter/google_sign_in.js (hunk 8 above)
Comment (rtitle) [reply src/main/resources/jupyter/google_sign_in.js]: 👍 
I'm used to immutable variables :(

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/SwaggerConfig.scala (hunk 9 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/SwaggerConfig.scala]: Yeah I suppose we are. I can make this `String` to be consistent.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/ClusterResourcesConfig.scala [hunk 10]
```diff
@@ -1,14 +1,14 @@
...
-                                   jupyterServerKey: String,
-                                   jupyterRootCaPem: String,
-                                   jupyterRootCaKey: String,
-                                   jupyterProxySiteConf: String,
-                                   jupyterInstallExtensionScript: String
-                                 )
\ No newline at end of file
+import org.broadinstitute.dsde.workbench.leonardo.model.ClusterResource
+
+case class ClusterResourcesConfig(initActionsScript: ClusterResource,
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/ClusterResourcesConfig.scala]: Super minor but is it necessary to a have `ClusterResource` case class, maybe a generic `FileName` would be better? idk which is superior honestly. Just musing.

Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/ClusterResourcesConfig.scala]: Well, I wanted this case class to specifically represent resources under `src/main/resources/jupyter` which should be uploaded to the cluster init bucket. Trying to distinguish between cluster "resources" (inside the leonardo.jar) and cluster "files" (on the filesystem of the leo docker).

**Pull request merged**